finetune/target_models/sequence_labeling.py

```diff
@@ -217,25 +275,75 @@ def finetune(self, Xs, Y=None, context=None, update_hook=None):
+            if self.config["unknown_labels"]:
+                model_copy.input_pipeline.target_dim -= 1
+                if self.config["unknown_token"] in model_copy.input_pipeline.label_encoder.target_labels:
+                    model_copy.input_pipeline.label_encoder.target_labels.remove(self.config["unknown_token"])
+
+            # Predict on full dataset
             initial_run_preds = []
             outer_batch_size = self.config.predict_batch_size
             for b_start in range(0, len(Xs), outer_batch_size):
-                initial_run_preds += model_copy.predict(Xs[b_start: b_start + outer_batch_size])
+                initial_run_preds += model_copy.predict(
+                    Xs[b_start : b_start + outer_batch_size]
+                )
             del model_copy
-            Y_with_neg_samples = negative_samples(initial_run_preds, Y, pad=self.config.pad_token)
+
+            # Tag negative predictions with <PAD> label and add to label set
+            Y_with_neg_samples = negative_samples(
+                initial_run_preds, Y, pad=self.config.pad_token
+            )
+
+            # If neg_samples_path is provided, write Y_with_neg_samples to path
+            if neg_samples_path:
+                with open(neg_samples_path, "w") as handle:
+                    json.dump(Y_with_neg_samples, handle, cls=NpEncoder)
+
             # this means we get the same absolute number of randomly sampled empty chunks with or without this option.
-            self.config.max_empty_chunk_ratio *= sum(len(yi) for yi in Y) / sum(len(yi) for yi in Y_with_neg_samples)
+            self.config.max_empty_chunk_ratio *= sum(len(yi) for yi in Y) / sum(
+                len(yi) for yi in Y_with_neg_samples
+            )
             Y = Y_with_neg_samples
 
             # Reinitialize the model including rebuilding the saver.
             self._initialize()
-        return super().finetune(Xs, Y=Y, context=context, update_hook=update_hook)
+
+            # If training with partial labels, combine output of auto negative sampling on
+            # gold data with partial labels data
+            if X_partial and Y_partial:
+                Xs = Xs + X_partial
+                Y = Y + Y_partial
+
+                # Set empty chunks ratio to 0 to remove empty chunks from partially labeled
+                # data, which will only have a sample of true labels
+                # TODO Determine if we need something more sophisticated for chunking
+                self.config.max_empty_chunk_ratio = 0.0
+
+        super().finetune(Xs, Y=Y, context=context, update_hook=update_hook)
+
+        # This is hacky and there's probably a better way to do this
+        # This is required because otherwise the tensor shapes do not match during prediction
+        unknown_token = self.config["unknown_token"]
+        if (
+            self.config["unknown_labels"]
+            and unknown_token in self.input_pipeline.label_encoder.target_labels
```

Comment:
Maybe some of these -1's could be handled by making more of this a responsibility of the label encoder?

Reply:
Yeah I like that idea. Basically I'm running into a scenario where I need the extra `<UNK>` dim whenever I train, but not when I predict. And the code gets hairier anytime there's multiple train/predict loops (like w/ ANS or partial labels creation)